File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/converter/FileValueConverter.java
package org.skellig.teststep.processing.converter;


import org.skellig.teststep.processing.exception.TestStepProcessingException;

import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class FileValueConverter implements TestStepValueConverter {

    private static final String FILE_REGEX = "file\\((.+)\\)";

    private final Pattern filePathPattern;
    private final ClassLoader classLoader;

    public FileValueConverter(ClassLoader classLoader) {
        this.classLoader = classLoader;
        filePathPattern = Pattern.compile(FILE_REGEX);
    }

    @Override
    public Object convert(String value) {
        Matcher matcher = filePathPattern.matcher(value);
        if (matcher.find()) {
            return readFileContentFromFilePath(matcher.group(1));
        }
        return value;
    }

    private String readFileContentFromFilePath(String pathToFile) {
        URL resource = classLoader.getResource(pathToFile);
        if (resource != null) {
            try {
                return new String(Files.readAllBytes(Paths.get(resource.toURI())), StandardCharsets.UTF_8);
            } catch (Exception e) {
                throw new TestStepProcessingException(String.format("Failed to read file '%s'", pathToFile), e);
            }
        } else {
            throw new TestStepProcessingException(String.format("File '%s' doesn't exist", pathToFile));
        }
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/converter/FileValueConverter.kt
package org.skellig.teststep.processing.converter

import org.skellig.teststep.processing.exception.TestDataConversionException
import org.skellig.teststep.processing.exception.TestValueConversionException
import org.skellig.teststep.processing.experiment.FunctionValueProcessor
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Paths
import java.util.regex.Pattern

class FileValueConverter(val classLoader: ClassLoader) : FunctionValueProcessor {

    override fun execute(name: String, args: Array<Any?>): Any =
        if (args.size == 1) {
            readFileContentFromFilePath(args[0]?.toString() ?: "")
        } else {
            throw TestDataConversionException("Function `fromFile` can only accept 1 String argument. Found ${args.size}")
        }

    private fun readFileContentFromFilePath(pathToFile: String): String {
        val resource = classLoader.getResource(pathToFile)
        return if (resource != null) {
            try {
                String(Files.readAllBytes(Paths.get(resource.toURI())), StandardCharsets.UTF_8)
            } catch (e: Exception) {
                throw TestValueConversionException(String.format("Failed to read file '%s'", pathToFile), e)
            }
        } else {
            throw TestValueConversionException(String.format("File '%s' doesn't exist", pathToFile))
        }
    }

    override fun getFunctionName(): String = "fromFile"
}